node/src/main/scala/org/bitcoins/node/Node.scala

```diff
@@ -44,24 +45,62 @@ trait Node extends NodeApi with ChainQueryApi with P2PLogger {
 
   implicit def executionContext: ExecutionContext = system.dispatcher
 
-  val peers: Vector[Peer]
+  def peers: Vector[Peer] = peerData.keys.toVector
 
-  private val _peerServices: mutable.Map[Peer, ServiceIdentifier] =
-    mutable.Map.empty
+  case class PeerData(peer: Peer) {
+
+    lazy val peerMessageSender: PeerMessageSender = PeerMessageSender(client)
+
+    lazy val client: P2PClient = {
+      val peerMessageReceiver =
+        PeerMessageReceiver.newReceiver(node = Node.this, peer = peer)
+      P2PClient(context = system,
+                peer = peer,
+                peerMessageReceiver = peerMessageReceiver)
+    }
+
+    private var _serviceIdentifier: ServiceIdentifier = _
```

Comment:
Don't do `_` when initializing a variable, it's equivalent to using `null` which leads to `NullPointerException`. Just do 

```scala
private var _serviceIdentifierOpt: Option[ServiceIdentifier] = None
```